File: graph_algorithms/week2_graph_decomposition2/2_order_of_courses/toposort.py
from typing import List


def dfs(adj: List[List[int]], used: List[bool], order: List[int], v: int) -> None:
    used[v] = True

    for w in adj[v]:
        if not used[w]:
            dfs(adj, used, order, w)

    order.append(v)


def toposort(adj: List[List[int]]) -> List[int]:
    used = [False] * len(adj)
    order: List[int] = []

    for v in range(len(adj)):
        if not used[v]:
            dfs(adj, used, order, v)

    return list(reversed(order))


if __name__ == "__main__":  # noqa: C901
    n, m = map(int, input().split())

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)

    order = toposort(adj)
    for x in order:
        print(x + 1, end=" ")
    print()

File: graph_algorithms/week2_directed_graphs/2_order_of_courses/toposort.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using std::pair;
using std::vector;

void dfs(vector<vector<int>> &adj, vector<bool> &used, vector<int> &order, int v) {
  used[v] = true;

  for (int w : adj[v])
    if (!used[w])
      dfs(adj, used, order, w);

  order.push_back(v);
}

vector<int> toposort(vector<vector<int>> &adj) {
  vector<bool> used = vector<bool>(adj.size(), false);
  vector<int> order;

  for (int v = 0; v < adj.size(); v++)
    if (!used[v])
      dfs(adj, used, order, v);

  std::reverse(order.begin(), order.end());

  return order;
}

int main() {
  size_t n, m;
  std::cin >> n >> m;

  vector<vector<int>> adj(n, vector<int>());
  for (size_t i = 0; i < m; i++) {
    int a, b;
    std::cin >> a >> b;
    adj[a - 1].push_back(b - 1);
  }

  vector<int> order = toposort(adj);
  for (size_t i = 0; i < order.size(); i++)
    std::cout << order[i] + 1 << " ";
  std::cout << std::endl;
}
